drivers/adc/adc_stm32.c

```diff
@@ -984,8 +1012,11 @@ static void adc_stm32_isr(const struct device *dev)
 	ADC_TypeDef *adc = config->base;
 
 	*data->buffer++ = LL_ADC_REG_ReadConversionData32(adc);
-
-	adc_context_on_sampling_done(&data->ctx, dev);
+	data->samples_count++;
+	if (data->samples_count == data->channel_count) {
+		adc_context_on_sampling_done(&data->ctx, dev);
+		adc_stm32_teardown_channels(dev);
```

Comment:
This seems unexpected to me.

If `adc_context_on_sampling_done()` calls the user call back and it returns `ADC_ACTION_REPEAT`, or the buffer is not finished because it has room for multiple samples, then `adc_context_on_sampling_done()` will cause `adc_context_start_sampling()` to get called again for another sample.

It seems odd that the channels get torn down here.  When `adc_context_start_sampling()` is called it will call `adc_stm32_start_conversion()` and that appears to re-enable the ADC.  It seems like the channels never get setup again.

Reply:
This seems to be an issue with the existing driver as well, where the channels are torn down but not configured again when `ADC_ACTION_REPEAT`. The existing driver might even teardown the channel before the read is done if in `async`.

The only way to teardown the channels at the correct time is to do it when `adc_context_complete`, but currently `adc_stm32.c` does not know when this occurs. We could add a `on_context_complete` callback that the stm32 driver can subscribe to?

Reply:
I think adc_context makes the assumption that if sampling is driven by a kernel timer, then it can just stop trigger when sampling is done.  There is no need to unconfigure the input mux or anything else.  So there is no hook into `adc_context_complete()`.

If the sampling is done by an ADC hardware timer (maybe a "continuous-mode" bit in the ADC), then it will need to be turned off when sampling is done.  But in this case the driver does not define `ADC_CONTEXT_USES_KERNEL_TIMER` and the adc_context code will give the driver a hook `adc_context_disable_timer()` where it can place code to run when the entire sequence is finished.

I can think of two ways to solve it:

1. I don't think the driver must teardown the channels.  It's not going to generate more interrupts.  So don't.  Move that feature to power management support.
2. It's reasonable that some ADCs need to do stuff other than turn off a timer when they are done and there should be hook in adc_context for this.  The style here would probably be a driver static function with a fixed `adc_context_*` name, rather than a function pointer, like how `adc_context_disable_timer()` is implemented.

Reply:
I would prefer leaving the teardown channels in, because it is useful low-power applications, as mentioned when it was added in https://github.com/zephyrproject-rtos/zephyr/commit/411079cf1d885747fcb9f84b27e168c09e895226.

I think we should go with the second option you list there, which is similar to my proposal except we're not storing a function pointer.

Reply:
> I think we should go with the second option you list there, which is similar to my proposal except we're not storing a function pointer.

@heinwessels If you do please think about doing it in a separate commit ;-)

Reply:
Slight variation on 2, have a macro, like the existing `ADC_CONTEXT_USES_KERNEL_TIMER`, that causes adc_context.h to not define `adc_context_complete()`.  The driver can do it.  It would have to duplicate the existing actions of that function, but it's not that much, and this way the driver would be able to decide if it should notify the thread waiting for the ADC data before or after it does the channel teardown, or other complications with how the sequence should finish.  So one exchanges code duplication for flexibility.

But whatever you decide, if you don't fix the driver in this PR, it would be nice to document that repeated sampling doesn't work, and check if `sequence->options->extra_samplings` is non-zero and return EINVAL.  No way to detect ADC_ACTION_REPEAT though.

Reply:
I've added a `__weak` function called `adc_context_on_complete` that can optionally be implemented by concrete drivers that is called from `adc_context_complete`. The `adc_stm32_teardown_channels` is then only called from that location.

Reply:
Although now I see a CI warning
```
 WEAK_DECLARATION: Using weak declarations can have unintended link defects
```
I'll do something similar to `ADC_CONTEXT_USES_KERNEL_TIMER` then.

Reply:
Okay, now it's a function enabled if `ADC_CONTEXT_ENABLE_ON_COMPLETE` is defined.